### ISSUE TITLE
2.0.12: numactrl isn't LTO ready

### ISSUE THREAD

**Issue by kloczek** [numactl/numactl#63]
Without LTO optimisation everything  is OK.
<pre>/bin/sh ./libtool  --tag=CC   --mode=link gcc -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -flto   -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -flto -fuse-linker-plugin -o numactl numactl.o util.o shm.o libnuma.la 
libtool: link: gcc -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -flto -Wl,-z -Wl,relro -Wl,--as-needed -Wl,-z -Wl,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -flto -fuse-linker-plugin -o .libs/numactl numactl.o util.o shm.o  ./.libs/libnuma.so
/usr/bin/ld: /tmp/numactl.ikwiBY.ltrans0.ltrans.o: in function `show_physcpubind&apos;:
/home/tkloczko/rpmbuild/BUILD/numactl-2.0.12/numactl.c:115: undefined reference to `numa_sched_getaffinity&apos;
/usr/bin/ld: /tmp/numactl.ikwiBY.ltrans0.ltrans.o: in function `main&apos;:
/home/tkloczko/rpmbuild/BUILD/numactl-2.0.12/numactl.c:140: undefined reference to `numa_get_run_node_mask&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/numactl-2.0.12/numactl.c:143: undefined reference to `numa_get_interleave_mask&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/numactl-2.0.12/numactl.c:144: undefined reference to `numa_get_membind&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/numactl-2.0.12/numactl.c:509: undefined reference to `numa_tonodemask_memory&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/numactl-2.0.12/numactl.c:449: undefined reference to `numa_interleave_memory&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/numactl-2.0.12/numactl.c:489: undefined reference to `numa_sched_setaffinity&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/numactl-2.0.12/numactl.c:228: undefined reference to `numa_node_to_cpus&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/numactl-2.0.12/numactl.c:511: undefined reference to `numa_set_membind&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/numactl-2.0.12/numactl.c:451: undefined reference to `numa_set_interleave_mask&apos;
collect2: error: ld returned 1 exit status
</pre>

**Comment by filbranden** [comment]
Can you give more details on how to reproduce this issue?

Please list your Linux distribution, compiler version, etc. Also, what options are you passing to `./configure` to request an LTO build? I imagine this is going through `CFLAGS` or `LDFLAGS`, but knowing the exact ones you're passing would be helpful.

If you have pointers on LTO and common issues with it that might be helpful in troubleshooting this, I'd appreciate it. I know a bit about LTO but not too much...

Thanks!
Filipe

**Comment by kloczek** [comment]
I'm not passing any options to configure parametrs. Automake/aytocof/lbtool offers completly trannsparrent way to alter optimisation by passing AR, NM,. RANLIB,  CC, CXX, CFALAGX, CXXFLAGS and LDFLAGS.
This what I',m passing in configure envinroment variables.

<pre>$ rpm -E %set_build_flags

	export CFLAGS=&quot;${CFLAGS:--O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -flto}&quot;; 
	export CXXFLAGS=&quot;${CXXFLAGS:--O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -flto}&quot;; 
	export FFLAGS=&quot;${FFLAGS:--O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -I/usr/lib64/gfortran/modules -flto}&quot;; 
	export FCFLAGS=&quot;${FCFLAGS:--O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -I/usr/lib64/gfortran/modules -flto}&quot;; 
	export LDFLAGS=&quot;${LDFLAGS:--Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -flto -fuse-linker-plugin}&quot;; 
	export AR=&quot;gcc-ar&quot; RANLIB=&quot;gcc-ranlib&quot; NM=&quot;gcc-nm&quot;; 
	export CC=&quot;gcc&quot; CXX=&quot;g++&quot; 
</pre>

**Comment by filbranden** [comment]
@kloczek More details about your environment would be appreciated... Did you configure your RPM to build with `-flto` yourself?

I'm on Fedora Rawhide (gcc-9.0.1-0.15.fc31, rpm-4.14.2.1-8.fc31) and my output of `rpm -E %set_build_flags` doesn't really include any `-flto` anywhere...

I'm trying to get a closer understanding of your build environment, so I can try to reproduce this as closely as your use case, to ensure a fix would work for you.

Cheers!
Filipe

**Comment by kloczek** [comment]
Yes I did and what l'm using is working for almost 500 other packages (which I've already tested).

**Comment by filbranden** [comment]
I managed to reproduce this by putting this in my `.rpmmacros`:

```
%build_cflags %{optflags} -flto
%build_cxxflags %{optflags} -flto
%build_ldflags -Wl,-z,relro %{_ld_as_needed_flags} %{_ld_symbols_flags} %{_hardened_ldflags} -flto

%set_build_flags \
  CFLAGS="${CFLAGS:-%{build_cflags}}" ; export CFLAGS ; \
  CXXFLAGS="${CXXFLAGS:-%{build_cxxflags}}" ; export CXXFLAGS ; \
  FFLAGS="${FFLAGS:-%{build_fflags}}" ; export FFLAGS ; \
  FCFLAGS="${FCFLAGS:-%{build_fflags}}" ; export FCFLAGS ; \
  LDFLAGS="${LDFLAGS:-%{build_ldflags}}" ; export LDFLAGS ; \
  AR="gcc-ar" ; export AR ; \
  RANLIB="gcc-ranlib" ; export RANLIB ; \
  NM="gcc-nm" ; export NM
```

Is this similar to the setup you are using? If you could share the details of the Linux distribution you are using, that would be helpful... As far as I can tell, Fedora is not really building *everything* with LTO (they just enabled it for Firefox.) Can you share *how* you configured your system to build using LTO? If you followed a guide or howto, can you share a link to the instructions?

It looks like the problem is when linking the library `libnuma.so` with `-flto`, it's getting rid of some symbols there. Marking those with `__attribute__((used))` might be enough, but I'm not sure whether that's the right thing to do or not... I'll make some experiments with that and post back here.

**Comment by kloczek** [comment]
Yes, that is very similar to what I have and I think that with what you have now should be able to reproduce that issue.
Yes, linked libnuma seems has much less symbols when it is linked with LTO so on first look it maybe related to symbols viability definition (sorry I've not been investigating that issue enough close to say more).

I'm using my own Linux distribution and temporary I cannot share exact details of my work (sorry). In my build env I'm using gcc 9 and sometimes clang however clang still needs to catch up with many things which are already correctly working with gcc.
I've not been testing yet numactl with clang and as I found that it fails with gcc chance that it will be working are close to null.

**Comment by kloczek** [comment]
BTW: POSIX sh allow to use "export foo=bar" instead "foo=bar; export foo" :)

**Comment by kloczek** [comment]
If you want I can share full rpmbuild log.

**Comment by kloczek** [comment]
Any progress on fixing that issue? :)

**Comment by kloczek** [comment]
Just found that new version has been released.
However looks like that version still is not LTO ready.
<pre>/usr/bin/ld: /tmp/numactl.56ZhSD.ltrans0.ltrans.o: in function `show_physcpubind&apos;:
/home/tkloczko/rpmbuild/BUILD/numactl-2.0.13/numactl.c:116: undefined reference to `numa_sched_getaffinity&apos;
/usr/bin/ld: /tmp/numactl.56ZhSD.ltrans0.ltrans.o: in function `main&apos;:
/home/tkloczko/rpmbuild/BUILD/numactl-2.0.13/numactl.c:141: undefined reference to `numa_get_run_node_mask&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/numactl-2.0.13/numactl.c:144: undefined reference to `numa_get_interleave_mask&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/numactl-2.0.13/numactl.c:145: undefined reference to `numa_get_membind&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/numactl-2.0.13/numactl.c:229: undefined reference to `numa_node_to_cpus&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/numactl-2.0.13/numactl.c:450: undefined reference to `numa_interleave_memory&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/numactl-2.0.13/numactl.c:452: undefined reference to `numa_set_interleave_mask&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/numactl-2.0.13/numactl.c:490: undefined reference to `numa_sched_setaffinity&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/numactl-2.0.13/numactl.c:510: undefined reference to `numa_tonodemask_memory&apos;
/usr/bin/ld: /home/tkloczko/rpmbuild/BUILD/numactl-2.0.13/numactl.c:512: undefined reference to `numa_set_membind&apos;
collect2: error: ld returned 1 exit status
</pre>

**Comment by filbranden** [comment]
Yes, unfortunately I didn't have time to figure this one out before releasing 2.0.13...

Hopefully, we'll have a solution for this one before the next release (and also hopefully we'll not have so much time in between releases!)

If you have any hints on how to fix this one, that would be very welcome, of course! I think I can manage to figure this one out eventually, but it might be trickier than I expect...

Cheers,
Filipe

**Comment by andikleen** [comment]
I looked at this now.

The problem is that libnuma uses symbol versions using inline assembler to use .symver. There doesn't seem to be a way to do the in gcc with function attributes. LTO doesn't understand the aliasing through symver.

However the basic problem is a build problem. The .a is supposed to be shipped as part of the package, and shouldn't be built with LTO anyways because that would make it incompatible with other compilers or other gcc versions.

Anyways I guess the easy fix is to add -fno-lto to libnuma, something like this. The only drawback I think is
that it cannot be easily overriden anymore in CFLAGS 

diff --git a/Makefile.am b/Makefile.am
index c71296094696..530ee6951bfd 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -39,6 +39,7 @@ migspeed_LDADD = libnuma.la -lrt
 memhog_SOURCES = memhog.c util.c
 memhog_LDADD = libnuma.la
 
+libnuma_la_CFLAGS = -g -O2 -fno-lto
 libnuma_la_SOURCES = libnuma.c syscall.c distance.c affinity.c affinity.h sysfs.c sysfs.h rtnetlink.c rtnetlink.h versions.ldscript
 libnuma_la_LDFLAGS = -version-info 1:0:0 -Wl,--version-script,$(srcdir)/versions.ldscript -Wl,-init,numa_init -Wl,-fini,numa_fini

**Comment by kloczek** [comment]
No offence but hardcoding exact compile options it is not how some issues should be solved. Please do not commit that patch.
Versioned symbols can be used with LTO.
https://stackoverflow.com/questions/46304742/how-to-combine-lto-with-symbol-versioning/

**Comment by filbranden** [comment]
This has been fixed with #92 and #93, at least on gcc 10 it's possible to build numactl with LTO. The version with the fixes has been released as 2.0.14.

I also found a few other places where CFLAGS were being overridden and I cleaned those up.

Let me know if you have any trouble with the build!

**Comment by kloczek** [comment]
Just checked ald all looks good :)

Thx.
